sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java

```diff
@@ -268,9 +309,9 @@ public void startBundle(StartBundleContext c) throws Exception {
     public void processElement(ProcessContext c) throws Exception {
       List<OutgoingMessage> pubsubMessages = new ArrayList<>(publishBatchSize);
       int bytes = 0;
-      for (OutgoingMessage message : c.element().getValue()) {
+      for (OutgoingMessage message : c.element()) {
         if (!pubsubMessages.isEmpty()
-            && bytes + message.message().getData().size() > publishBatchBytes) {
+            && bytes + message.getMessage().getData().size() > publishBatchBytes) {
```

Comment:
another option would be to change so we group by topic+shard instead of just shard
that would keep this simple here, and avoid sequential blocking publishes

Reply:
Added grouping by topic+shard